manticore/core/executor.py

```diff
@@ -98,164 +84,243 @@ def __str__(self):
+
+        #Number of currently running workers. Initially no runnign workers
         self._running = manager.Value('i', 0 )
-        self._count = manager.Value('i', 0 )
-        self._shutdown = manager.Value('i', 0)
-        self._visited = manager.list()
-        self._errors = manager.list()
-        self._all_branches = manager.list()
-        self._lock = manager.Condition(manager.Lock())
-        #Stats
-        self._stats = manager.list()
-
-        #search previously generated states
-        saved_states = [ self._getFilename(filename) for filename in os.listdir(self.workspace) if filename.endswith('.pkl') ]
+
+        #Number of generated testcases
+        self._test_count = manager.Value('i', 0 )
+
+        #Number of total intermediate states
+        self._state_count = manager.Value('i', 0 )
+
+        #Executor wide shared context
+        if context is None:
+            context = {}
+        self._shared_context = manager.dict(context)
+    
+        #scheduling priority policy (wip)
+        self.policy = Random()
+
+        if self.load_workspace():
+            if initial is not None:
+                logger.error("Ignoring initial state")
+        else:
+            if initial is not None:
+                self.add(initial)
+                ##FIXME PUBSUB  We need to forward signals here so they get declared
+                ##forward signals from initial state so they are declared here
+                forward_signals(self, initial, True)
+
+    @contextmanager
+    def locked_context(self):
+        ''' Executor context is a shared memory object. All workers share this. 
+            It needs a lock. Its used like this:
+
+            with executor.context() as context:
+                vsited = context['visited']
+                visited.append(state.cpu.PC)
+                context['visited'] = visited
+        '''
+        with self._lock:
+            yield self._shared_context
+
+
+
+    def _register_state_callbacks(self, state, state_id):
+        '''
+            Install forwarding callbacks in state so the events can go up. 
+            Going up, we prepend state in the arguments.
+        ''' 
+        #Forward all state signals
+        forward_signals(self, state, True)
+
+    def add(self, state):
+        '''
+            Enqueue state.
+            Save state on storage, assigns an id to it, then add it to the 
+            priority queue
+        '''
+        #save the state to secondary storage
+        state_id = self.store(state)
+        #add the state to the list of pending states
+        self.put(state_id)
+        return state_id
+
+    def load_workspace(self):
+        #Browse and load states in a workspace in case we are trying to 
+        # continue from paused run
+        saved_states = []
+        for filename in os.listdir(self.workspace):
+            if filename.startswith('state_') and filename.endswith('.pkl'):
+                saved_states.append(self._getFilename(filename)) 
```

Comment:
can we add a regression test for these small bugs? even the most basic test of this function would have failed because the bugs were very simple (nonexistent functions/variables)